dapp/contracts/Crowdsale.sol

```diff
@@ -0,0 +1,210 @@
+        if (approved(_address)) {
+            return;
+        }
+
+        ValidationRequest(this, _address);
+        bytes32 id = boson().queryAddress(_address);
+        validationRequests[id] = _address;
+    }
+
+    function __bosonCallback(bytes32 _id, bool _status) public {
+        require(msg.sender == boson().cbAddress());
+        address requested = validationRequests[_id];
+        delete validationRequests[_id];
+        ValidationResult(this, requested, _status);
+        if (_status) {
+            validationResults[requested].approvals++;
+            validationResults[requested].rejections = 0;
+        } else {
+            validationResults[requested].rejections++;
+        }
+    }
+
+    function boson() private returns (IBoson _boson) {
+        _boson = IBoson(bosonResolver.getAddress());
+    }
+
+}
+
+contract Crowdsale is Ownable, BosonApproved {
+    uint256 public GOAL_WEI = 10000 ether;
+    uint256 public DURATION = 14 days;
+    uint256 public MINIMUM_PURCHASE_WEI = 1 ether / 100;
+
+    using SafeMath for uint256;
+
+    enum State {
+        Initialized,
+        Started,
+        Finished
+    }
+
+    State public state;
+    GetToken public token;
+
+    uint256 public tokensLeft;
+    uint256 public tokensGoal;
+    uint256 public startDate;
+    uint256 public endDate;
+
+    event StateTransition(uint256 _oldState, uint256 _newState);
+    event StateTimeout(uint256 _oldState);
+
+    function nextState(State _newState) private {
+        StateTransition(uint256(state), uint256(_newState));
+        state = _newState;
+    }
+
+    function processTimeouts() private {
+        if (state == State.Started) {
+            if (block.timestamp > endDate) {
+                StateTimeout(uint256(state));
+                nextState(State.Finished);
+            }
+        }
+    }
+
+    function Crowdsale(GetToken _token, IBosonAddressResolver _bar) public {
+        token = _token;
+        bosonResolver = _bar;
+        owner = msg.sender;
+        state = State.Initialized;
+    }
+
+    function start() onlyOwner public {
+        require(state == State.Initialized);
+
+        state = State.Started;
+        startDate = block.timestamp;
+        endDate = startDate + DURATION;
+        tokensGoal = token.balanceOf(this);
+        tokensLeft = tokensGoal;
+
+        require(tokensGoal > 0);
+        require(price() > 0);
+    }
+
+    function price() public view returns (uint256 _price) {
+        require(state >= State.Started);
+        return GOAL_WEI / tokensGoal;
+    }
+
+    function buy(address buyer) public payable {
+        processTimeouts();
+        require(state == State.Started);
+        require(buyer != 0x0);
+        require(approved(buyer));
+
+        // Convert purchase to requested tokens.
+        uint256 tokens = msg.value.div(price());
+
+        // Make sure to not sell more than we have.
+        if (tokensLeft < tokens) {
+            tokens = tokensLeft;
```

Comment:
what happens with the rest of the money entered?

Reply:
Fixed, added test.